### PR TITLE
Don't use editable mode in tox

### LINKED ISSUE #20: Running the `py` tox environment twice in a row fails
```
$ pip --version
pip 21.3.1 from /home/babolivier/.virtualenvs/synapse-py39/lib/python3.9/site-packages/pip (python 3.9)
$ tox -e py
py create: /home/babolivier/Documents/matrix/ems-synapse-background-update-controller/.tox/py
py installdeps: git+https://github.com/matrix-org/synapse.git@develop#egg=matrix-synapse
py develop-inst: /home/babolivier/Documents/matrix/ems-synapse-background-update-controller
py installed: aiounittest==1.4.1,attrs==21.2.0,Automat==20.2.0,backports.entry-points-selectable==1.1.1,bcrypt==3.2.0,black==21.9b0,bleach==4.1.0,boto3==1.20.17,boto3-stubs==1.20.17,botocore==1.23.17,botocore-stubs==1.23.17,canonicaljson==1.5.0,certifi==2021.10.8,cffi==1.15.0,charset-normalizer==2.0.8,click==8.0.3,constantly==15.1.0,cryptography==36.0.0,distlib==0.3.3,-e git+https://gitlab.matrix.org/new-vector/ems-synapse-background-update-controller.git@832a4895ede04ec4edc39fc94a9bd43e6a8007a3#egg=ems_background_update_controller,filelock==3.4.0,flake8==4.0.1,frozendict==2.1.1,hyperlink==21.0.0,idna==3.3,ijson==3.1.4,importlib-metadata==4.8.2,incremental==21.3.0,isort==5.9.3,Jinja2==3.0.3,jmespath==0.10.0,jsonschema==4.2.1,MarkupSafe==2.0.1,matrix-synapse @ git+https://github.com/matrix-org/synapse.git@153194c7717d8016b0eb974c81b1baee7dc1917d,mccabe==0.6.1,msgpack==1.0.3,mypy==0.910,mypy-boto3-cloudwatch==1.20.9,mypy-extensions==0.4.3,netaddr==0.8.0,packaging==21.3,pathspec==0.9.0,phonenumbers==8.12.38,Pillow==8.4.0,platformdirs==2.4.0,pluggy==1.0.0,prometheus-client==0.12.0,py==1.11.0,pyasn1==0.4.8,pyasn1-modules==0.2.8,pycodestyle==2.8.0,pycparser==2.21,pyflakes==2.4.0,pymacaroons==0.13.0,PyNaCl==1.4.0,pyOpenSSL==21.0.0,pyparsing==3.0.6,pyrsistent==0.18.0,python-dateutil==2.8.2,PyYAML==6.0,regex==2021.11.10,requests==2.26.0,s3transfer==0.5.0,service-identity==21.1.0,signedjson==1.1.1,simplejson==3.17.6,six==1.16.0,sortedcontainers==2.4.0,toml==0.10.2,tomli==1.2.2,tox==3.24.4,treq==21.5.0,Twisted==21.7.0,typing_extensions==4.0.1,unpaddedbase64==2.1.0,urllib3==1.26.7,virtualenv==20.10.0,webencodings==0.5.1,wrapt==1.13.3,zipp==3.6.0,zope.interface==5.4.0
py run-test-pre: PYTHONHASHSEED='806511979'
py run-test: commands[0] | python -m unittest discover
....
----------------------------------------------------------------------
Ran 4 tests in 0.009s

OK
___________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________
  py: commands succeeded
  congratulations :)
$ tox -e py
/home/babolivier/Documents/matrix/ems-synapse-background-update-controller/.tox/py/bin/python: can't open file '/home/babolivier/Documents/matrix/ems-synapse-background-update-controller/setup.py': [Errno 2] No such file or directory
ERROR: invocation failed (exit code 2)
___________________________________________________________________________________________________________________ summary ___________________________________________________________________________________________________________________
ERROR:   py: InvocationError for command /home/babolivier/Documents/matrix/ems-synapse-background-update-controller/.tox/py/bin/python /home/babolivier/Documents/matrix/ems-synapse-background-update-controller/setup.py --name (exited with code 2)
```

The workaround that doesn't involve changing `tox.ini` seems to be deleting `.tox` before each run.

In https://github.com/matrix-org/synapse-module-cookiecutter-template/pull/18 I found that removing the `usedevelop=True` line made it pass, this seems to be corroborated by https://github.com/tox-dev/tox/issues/2197

According to @reivilibre (https://github.com/matrix-org/synapse-module-cookiecutter-template/pull/18#issuecomment-983951423) we should be able to work around this issue by dropping the `usedevelop=True` and instead add the `--editable` flag to the `deps` line (and replacing `extras` with `dev`).

### PR DESCRIPTION
**Pull request by babolivier**
Fixes #20

### CODE REVIEW

**Comment by callahad** [comment]
Related conversation in #18

**Comment by callahad** [comment]
I'm curious if not using editable installs at all is possible? Sounds like avoiding editable installs prevents whole classes of potential packaging bugs.

**Comment by babolivier** [comment]
> I'm curious if not using editable installs at all is possible? Sounds like avoiding editable installs prevents whole classes of potential packaging bugs.

The comment says it's needed for trial, but I managed to have it run just fine with only this config for the `py` environment:

```ini
[testenv:py]

extras = dev

commands =
  python -m twisted.trial tests
```

@reivilibre do you remember what lead to [this conclusion](https://github.com/matrix-org/synapse-module-cookiecutter-template/blob/main/%7B%7B%20cookiecutter.directory_name%20%7D%7D/tox.ini#L9-L15)?

**Comment by callahad** [comment]
Looks like it's from this commit four years ago https://github.com/matrix-org/synapse/commit/4dd61df6f8d8d622b1327e2ce678d26e9c6911b0

**Comment by reivilibre** [comment]
Yeah, I copied it from Synapse because:

* consistency
* it sounded logical and reasonable
* I've had issues in the past with Trial and where it looks for tests to run.

The reason it seems to work here is because of the `-m` flag to Python, which afaik adds modules in the same directory to the module path as a side-effect of what it does. If you replace `python -m twisted.trial` with just `trial` you'll find it doesn't work.

That doesn't concern me however, since it's Tox and we can just use `python -m twisted.trial`. So I'm happy for you to remove editable mode if you find that helps things.

You **will** notice that using non-editable mode will be slower because it has to reinstall the package (but not necessarily the deps) every time, whereas in editable mode this is a no-op. However, editable mode is currently broken so the point is moot...

(unless that meant we preferred a `import setuptools; setuptools.setup()` `setup.py` file for compatibility — I'm not super fussed about which way you choose to do it as I don't habitually run through tox anyway... :P)

**Comment by babolivier** [comment]
Right, let's drop the editable mode then. We're talking about modules here anyway, which are usually fairly small, so the time save introduced by using the editable mode (even if `usedevelop` worked) probably doesn't outweigh the additional reliability in packaging not using it brings.

**Comment by richvdh** [comment]
> The reason it seems to work here is because of the -m flag to Python, which afaik adds modules in the same directory to the module path as a side-effect of what it does. 

the same directory as... what?

**Comment by reivilibre** [comment]
> the same directory as... what?

the present working directory.

```
rei@lithium ┄┄ /tmp $ python -m ptpython
>>> import sys
>>> sys.path
['', '/tmp', '/usr/lib/python39.zip', '/usr/lib/python3.9', '/usr/lib/python3.9/lib-dynload', '/home/rei/.local/lib/python3.9/site-packages', '/home/rei/work/PS/synapse-my-module', '/usr/local/lib/python3.9/dist-packages', '/usr/lib/python3/dist-packages', '/usr/lib/python3.9/dist-packages']

>>>                                                                                                                                                                                                                                                                                                                          

rei@lithium ┄┄ /tmp $ ptpython
>>> import sys
>>> sys.path
['', '/home/rei/.local/bin', '/usr/lib/python39.zip', '/usr/lib/python3.9', '/usr/lib/python3.9/lib-dynload', '/home/rei/.local/lib/python3.9/site-packages', '/home/rei/work/PS/synapse-my-module', '/usr/local/lib/python3.9/dist-packages', '/usr/lib/python3/dist-packages', '/usr/lib/python3.9/dist-packages']

>>>
```

Can't promise I'm not blaming the wrong thing — but all tools that I use that interact with the path seem to be affected this way (which has really confused me in the past until I noticed).

**Comment by richvdh** [comment]
> the present working directory.

how does that help it find the tests? Doesn't `tox` chdir to the right `.tox` directory?

**Comment by reivilibre** [comment]
> > the present working directory.
> 
> how does that help it find the tests? Doesn't `tox` chdir to the right `.tox` directory?

No... (or at least not in this case? I imagine you must have some reason for thinking it might!)

Reproduction:
```bash
rei@lithium ┄┄ ~/work/PS/synapse-my-module $ cat tests/__init__.py 
import os
print("cwd ", os.getcwd())
```

then

```bash
rei@lithium ┄┄ ~/work/PS/synapse-my-module $ tox -e py
(snip)

py run-test: commands[0] | python -m twisted.trial tests
cwd  /home/rei/work/PS/synapse-my-module 

(snip)
```

**Review by squahtx — APPROVED** [review]
No objections here!

I found the discussion hard to follow. The summary appears to be:
#18 happens when we use `usedevelop=True`, since that option expects a `setup.py`. Removing it makes tox work twice in a row, as noted in the issue.
We originally added `usedevelop=True` in Synapse as a workaround [[link]](https://github.com/matrix-org/synapse-module-cookiecutter-template/pull/21#issuecomment-984609953) which does not appear to be necessary for how we invoke trial. [[link]](https://github.com/matrix-org/synapse-module-cookiecutter-template/pull/21#issuecomment-984631250)
